File: find_order.dart
class Solution {
  List<List<int>> edges = [];
  List<int> visited = [];
  List<int> result = [];
  bool valid = true;

  List<int> findOrder(int numCourses, List<List<int>> prerequisites) {
    edges = List.generate(numCourses, (index) => []);
    visited = List.generate(numCourses, (index) => 0);
    result = [];
    valid = true;
    for (var prerequisite in prerequisites) {
      edges[prerequisite[1]].add(prerequisite[0]);
    }
    for (int i = 0; i < numCourses && valid; i++) {
      if (visited[i] == 0) {
        dfs(i);
      }
    }
    return valid == true ? result.reversed.toList() : [];
  }

  void dfs(int u) {
    visited[u] = 1;
    for (int v in edges[u]) {
      if (visited[v] == 0) {
        dfs(v);
        if (!valid) {
          return;
        }
      } else if (visited[v] == 1) {
        valid = false;
        return;
      }
    }
    visited[u] = 2;
    result.add(u);
  }
}

File: find_order.py
class Solution:
    def __init__(self):
        self.edges = []
        self.visited = []
        self.result = []
        self.valid = True
    
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.edges = [[] for _ in range(numCourses)]
        self.visited = [0 for _ in range(numCourses)]
        self.result = []
        self.valid = True

        for info in prerequisites:
            self.edges[info[1]].append(info[0])

        for i in range(numCourses):
            if self.valid:
                if self.visited[i] == 0:
                    self.dfs(i)
            else:
                break

        return self.valid == True and self.result or []

    def dfs(self, u):
        self.visited[u] = 1
        for v in self.edges[u]:
            if self.visited[v] == 0:
                self.dfs(v)
                if not self.valid:
                    return
            elif self.visited[v] == 1:
                self.valid = False
                return
        self.visited[u] = 2
        self.result.insert(0, u)